### PR TITLE
Feature/consensus context per spec

### PR DESCRIPTION
**Pull request by IdoZilberberg**
Impl and unit+component tests for RequestOrderingCommittee() and its support functions.
This is WIP, the algo for randomly choosing committee members is not final, see https://github.com/orbs-network/orbs-spec/issues/111

### CODE REVIEW

File: services/consensuscontext/test/consensus_context_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		if err != nil {
+			t.Error(err)
+		}
+		actualFederationSize := len(output.NodePublicKeys)
+		require.Equal(t, federationSize, actualFederationSize, "expected committee size is %d but got %d", federationSize, actualFederationSize)
+	})
+}
+
+// cc implements this
+func TestRequestValidationCommittee(t *testing.T) {
```
Comment (electricmonk) [inline services/consensuscontext/test/consensus_context_test.go]: eh?

File: services/consensuscontext/test/harness.go [hunk 2]
```diff
@@ -9,13 +9,27 @@ import (
...
 	"github.com/orbs-network/orbs-network-go/test/builders"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
 	"github.com/orbs-network/orbs-spec/types/go/protocol"
 	"github.com/orbs-network/orbs-spec/types/go/services"
 	"github.com/stretchr/testify/require"
 	"os"
 	"testing"
 )
 
+var federationNodePublicKeysForTest = []primitives.Ed25519PublicKey{
```
Comment (electricmonk) [inline services/consensuscontext/test/harness.go]: we have a `KeyPairForTest` function somewhere - are you sure you need this?

File: services/consensuscontext/committee.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	if requestedCommitteeSize > federationSize {
+		return federationSize
+	}
+	return requestedCommitteeSize
+}
+
+// Smart algo!
+func ChooseRandomCommitteeIndices(committeeSize uint32, randomSeed uint64) []uint32 {
```
Comment (electricmonk) [inline services/consensuscontext/committee.go]: why is it exported?

File: services/consensuscontext/committee.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	return keys
+}
+
+// TODO Pending a different impl if necessary
+func (s *service) RequestValidationCommittee(ctx context.Context, input *services.RequestCommitteeInput) (*services.RequestCommitteeOutput, error) {
+	return s.RequestOrderingCommittee(ctx, input)
+}
+
+func CalculateCommitteeSize(requestedCommitteeSize uint32, minimumCommitteeSize uint32, federationSize uint32) uint32 {
```
Comment (electricmonk) [inline services/consensuscontext/committee.go]: why is it exported?

File: services/consensuscontext/test/consensus_context_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+
+*/
+
+func TestRequestOrderingCommittee(t *testing.T) {
+	h := newHarness()
+	blockHeight := primitives.BlockHeight(1)
+	federationSize := len(h.config.FederationNodes(uint64(blockHeight)))
+
+	t.Run("if committee size <= federation size, then return same count of committee members as requested by config", func(t *testing.T) {
```
Comment (electricmonk) [inline services/consensuscontext/test/consensus_context_test.go]: why do you use this nesting? you're only sharing the harness and two simple variables between tests

File: services/consensuscontext/test/consensus_context_test.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
+package test
```
Comment (electricmonk) [inline services/consensuscontext/test/consensus_context_test.go]: Note that `RequestNewTransactionsBlock` and `RequestNewResultsBlock` are not implemented properly, you need to write component tests for them and make sure they are on par with spec

File: services/consensuscontext/test/harness.go (hunk 2 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/harness.go]: Comments for myself, removed.

File: services/consensuscontext/committee.go (hunk 4 above)
Comment (IdoZilberberg) [reply services/consensuscontext/committee.go]: Right, made it private. Leftover from when its tests were on "test" package.

File: services/consensuscontext/committee.go (hunk 3 above)
Comment (IdoZilberberg) [reply services/consensuscontext/committee.go]: Right, made it private. Leftover from when its tests were on "test" package.

File: services/consensuscontext/test/consensus_context_test.go (hunk 5 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/consensus_context_test.go]: I find this hierarchical construct clearer than the flat one, where harness init is repeated every test. Also, shared data between tests and overall number of tests in this suite will likely increase.

File: services/consensuscontext/test/consensus_context_test.go (hunk 1 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/consensus_context_test.go]: Leftovers with comments for myself. Removed.

File: services/consensuscontext/test/harness.go (hunk 2 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/harness.go]: The test needs an array of public keys, not a single keypair, so added a func to `keys.ed25519.go`.

File: services/consensuscontext/test/consensus_context_test.go (hunk 6 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/consensus_context_test.go]: They are old code I wrote with Tal in August. They are next in line for rewrite. This PR doesn't cover all of consensus context. Thinking of it, the PR name is misleading.

File: services/consensuscontext/test/harness.go (hunk 2 above)
Comment (IdoZilberberg) [reply services/consensuscontext/test/harness.go]: Refactored all this part out of existence.

**Comment by IdoZilberberg** [comment]
@electricmonk See if you have further comments. Once this is merge I'll continue with integration test.

**Pull request merged**